### PR TITLE
throw runtime `TypeError` on invalid ccall symbol

### LINKED ISSUE #45187: ccall with an invalid function as library gives internal error
```jl
julia> lib() = 1
lib (generic function with 1 method)

julia> ccall((:f, lib), Cvoid, ())
Internal error: encountered unexpected error during compilation of top-level scope:
TypeError(func=:ccall, context="", expected=Symbol, got=Main.lib)
jl_type_error_rt at /buildworker/worker/package_linux64/build/src/rtutils.c:119
jl_type_error at /buildworker/worker/package_linux64/build/src/rtutils.c:127
interpret_symbol_arg at /buildworker/worker/package_linux64/build/src/ccall.cpp:599
emit_ccall at /buildworker/worker/package_linux64/build/src/ccall.cpp:1253
emit_expr at /buildworker/worker/package_linux64/build/src/codegen.cpp:4610
emit_ssaval_assign at /buildworker/worker/package_linux64/build/src/codegen.cpp:4245
emit_stmtpos at /buildworker/worker/package_linux64/build/src/codegen.cpp:4487 [inlined]
...
ERROR: unsupported or misplaced expression foreigncall
Stacktrace:
 [1] Toplevel MethodInstance thunk
   @ Main REPL[2]:1
```

Should maybe give a better error?

### LINKED ISSUE #49141: Compiler error with unresolvable pointer for `ccall` in union split
The following code:

```julia
ccall(nothing, Cvoid, ())
```

produces a compiler error which leads to the expression being interpreted instead. This seems ok since the expression is not really a valid ccall. The error message is not really readable as in #45187. The message is caused by the following type check:

https://github.com/JuliaLang/julia/blob/3b2e0d8fbc10128c0f6f88823ccf6c33d5735abc/src/ccall.cpp#L663

However, the same error message can be produced in situations where the compiled ccall will never be executed such as a union split (not really sure about the term here). See the following code:

```julia
import Libdl

const libc = Libdl.dlopen("libc.so.6")

@inline call_printf(s, val) = ccall(s, Cint, (Cstring,), val)

function reproducer()
    s = Libdl.dlsym(libc, :printf) # ::Union{Nothing,Ptr{Cvoid}}
    call_printf(s, "hello")
end
```

In this case the inferred return type of `Libdl.dlsym` is `Union{Nothing,Ptr{Cvoid}}` and the `@code_typed` seems to indicate that the codegen for `call_printf(s,"hello")` was split in two branch (one for `s::Nothing` and one for `s::Ptr{Cvoid}`) leading the `s::Nothing` branch to throw the same compiler error when it is not executed at runtime (`s` being `Ptr{Cvoid}`):

```
julia> @code_typed reproducer()
# [...]
# 6 ── %18 = π (%15, Nothing)
# [...] 
# 17 ─ %46 = $(Expr(:foreigncall, :(%18), Int32, svec(Cstring), 0, :(:ccall), :(%44), "hello"))::Int32
```

> originally posted in https://github.com/fonsp/Pluto.jl/issues/2511

### PR DESCRIPTION
**Pull request by Pangoraw**
The `JL_TYPECHKS` in `interpret_symbol_arg` lead to compiler error message even in cases where the code will not be evaluated (see #45187, #49141). This PR replaces these compile-time checks with simply throwing a runtime `TypeError`. 

Comparison of error message given a ccall:

##### 1.9-rc1

<details>

```julia
julia> ccall(nothing, Cvoid, ())
Internal error: encountered unexpected error during compilation of top-level scope:
TypeError(func=:ccall, context="", expected=Ptr{T} where T, got=nothing)
ijl_type_error_rt at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:119
ijl_type_error at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:127
interpret_symbol_arg at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:663
emit_ccall at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:1349
emit_expr at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5276
emit_ssaval_assign at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:4860
emit_stmtpos at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5095 [inlined]
emit_function at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8073
jl_emit_code at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8408
jl_emit_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8456
_jl_compile_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:204
jl_generate_fptr_impl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:460
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2321 [inlined]
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2210
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2723 [inlined]
ijl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2738
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:903
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:856
ijl_toplevel_eval_in at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
eval_user_input at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:153
repl_backend_loop at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:249
#start_repl_backend#46 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:234
start_repl_backend at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:231
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#run_repl#59 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:377
run_repl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:363
jfptr_run_repl_60577.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#1017 at ./client.jl:421
jfptr_YY.1017_27809.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
jl_f__call_latest at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
run_main_repl at ./client.jl:405
exec_options at ./client.jl:322
_start at ./client.jl:522
jfptr__start_33350.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
true_main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:573
jl_repl_entrypoint at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:717
main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/cli/loader_exe.c:59
unknown function (ip: 0x7f73174fc78f)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
Internal error: encountered unexpected error during compilation of top-level scope:
TypeError(func=:ccall, context="", expected=Ptr{T} where T, got=nothing)
ijl_type_error_rt at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:119
ijl_type_error at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:127
interpret_symbol_arg at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:663
emit_ccall at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:1349
emit_expr at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5276
emit_ssaval_assign at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:4860
emit_stmtpos at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5095 [inlined]
emit_function at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8073
jl_emit_code at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8408
jl_emit_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8456
_jl_compile_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:204
jl_generate_fptr_for_unspecialized_impl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:512
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2335 [inlined]
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2210
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2723 [inlined]
ijl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2738
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:903
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:856
ijl_toplevel_eval_in at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
eval_user_input at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:153
repl_backend_loop at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:249
#start_repl_backend#46 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:234
start_repl_backend at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:231
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#run_repl#59 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:377
run_repl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:363
jfptr_run_repl_60577.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#1017 at ./client.jl:421
jfptr_YY.1017_27809.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
jl_f__call_latest at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
run_main_repl at ./client.jl:405
exec_options at ./client.jl:322
_start at ./client.jl:522
jfptr__start_33350.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
true_main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:573
jl_repl_entrypoint at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:717
main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/cli/loader_exe.c:59
unknown function (ip: 0x7f73174fc78f)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
ERROR: `ccall` requires the compiler
Stacktrace:
 [1] Toplevel MethodInstance thunk
   @ Main REPL[1]:1
```

</details>

##### This pr
```julia
julia> ccall(nothing, Cvoid, ())
ERROR: TypeError: in ccall: first argument not a pointer or valid constant expression, expected Ptr, got a value of type Nothing
Stacktrace:
 [1] top-level scope
   @ ./REPL[1]:1
```

##### 1.9-rc1

<details>

```julia
julia> ccall((:ok, lib), Cvoid, ())
Internal error: encountered unexpected error during compilation of top-level scope:
TypeError(func=:ccall, context="", expected=Symbol, got=Main.lib)
ijl_type_error_rt at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:119
ijl_type_error at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:127
interpret_symbol_arg at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:660
emit_ccall at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:1349
emit_expr at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5276
emit_ssaval_assign at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:4860
emit_stmtpos at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5095 [inlined]
emit_function at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8073
jl_emit_code at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8408
jl_emit_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8456
_jl_compile_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:204
jl_generate_fptr_impl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:460
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2321 [inlined]
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2210
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2723 [inlined]
ijl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2738
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:903
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:856
ijl_toplevel_eval_in at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
eval_user_input at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:153
repl_backend_loop at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:249
#start_repl_backend#46 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:234
start_repl_backend at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:231
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#run_repl#59 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:377
run_repl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:363
jfptr_run_repl_60577.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#1017 at ./client.jl:421
jfptr_YY.1017_27809.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
jl_f__call_latest at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
run_main_repl at ./client.jl:405
exec_options at ./client.jl:322
_start at ./client.jl:522
jfptr__start_33350.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
true_main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:573
jl_repl_entrypoint at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:717
main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/cli/loader_exe.c:59
unknown function (ip: 0x7f73174fc78f)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
Internal error: encountered unexpected error during compilation of top-level scope:
TypeError(func=:ccall, context="", expected=Symbol, got=Main.lib)
ijl_type_error_rt at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:119
ijl_type_error at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/rtutils.c:127
interpret_symbol_arg at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:660
emit_ccall at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/ccall.cpp:1349
emit_expr at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5276
emit_ssaval_assign at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:4860
emit_stmtpos at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:5095 [inlined]
emit_function at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8073
jl_emit_code at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8408
jl_emit_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/codegen.cpp:8456
_jl_compile_codeinst at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:204
jl_generate_fptr_for_unspecialized_impl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jitlayers.cpp:512
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2335 [inlined]
jl_compile_method_internal at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2210
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2723 [inlined]
ijl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2738
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:903
jl_toplevel_eval_flex at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:856
ijl_toplevel_eval_in at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
eval_user_input at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:153
repl_backend_loop at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:249
#start_repl_backend#46 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:234
start_repl_backend at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:231
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#run_repl#59 at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:377
run_repl at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:363
jfptr_run_repl_60577.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
#1017 at ./client.jl:421
jfptr_YY.1017_27809.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
jl_f__call_latest at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
run_main_repl at ./client.jl:405
exec_options at ./client.jl:322
_start at ./client.jl:522
jfptr__start_33350.clone_1 at /home/paul/.julia/juliaup/julia-1.9.0-rc1+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2731 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/gf.c:2913
jl_apply at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/julia.h:1878 [inlined]
true_main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:573
jl_repl_entrypoint at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/src/jlapi.c:717
main at /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/cli/loader_exe.c:59
unknown function (ip: 0x7f73174fc78f)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
ERROR: `ccall` requires the compiler
Stacktrace:
 [1] Toplevel MethodInstance thunk
   @ Main REPL[2]:1
```

</details>

##### This pr
```julia
julia> ccall((:ok, lib), Cvoid, ())
ERROR: TypeError: in ccall: first argument not a pointer or valid constant expression, expected Ptr, got a value of type Tuple{Symbol, typeof(lib)}
Stacktrace:
 [1] top-level scope
   @ ./REPL[4]:1
```

Most importantly, this removes the warning for https://github.com/JuliaLang/julia/issues/49141 since the `::Nothing` branch is never executed at runtime.

Closes #49141
Closes #45187

Feel free to close if this should be handled elsewhere/differently.

### CODE REVIEW

File: src/ccall.cpp [hunk 1]
```diff
@@ -696,7 +696,7 @@ static jl_cgval_t emit_cglobal(jl_codectx_t &ctx, jl_value_t **args, size_t narg
     Type *lrt = ctx.types().T_size;
     assert(lrt == julia_type_to_llvm(ctx, rt));
 
-    interpret_symbol_arg(ctx, sym, args[1], "cglobal", false);
+    interpret_symbol_arg(ctx, sym, args[1], /*ccall=*/false, false);
```
[the remark this answers is not in the document]
Comment (Pangoraw) [reply src/ccall.cpp:699]: I added error emission for `cglobal` too and added tests.

**Comment by vtjnash** [comment]
Looks like this cglobal test is hitting the interpreter and crashing instead in src/runtime_intrinsics.c

**Comment by Pangoraw** [comment]
The test has been fixed to trigger compilation instead of interpretation, the failing REPL test does not seem related to this pr.

**Comment by vtjnash** [comment]
Here is a fix for the runtime_intrinsics problem. Do you want to include this here too:
```diff
diff --git a/src/runtime_intrinsics.c b/src/runtime_intrinsics.c
index 9536ed1f02..99a81974a6 100644
--- a/src/runtime_intrinsics.c
+++ b/src/runtime_intrinsics.c
@@ -490,14 +490,14 @@ JL_DLLEXPORT jl_value_t *jl_cglobal(jl_value_t *v, jl_value_t *ty)
 
     char *f_lib = NULL;
     if (jl_is_tuple(v) && jl_nfields(v) > 1) {
-        jl_value_t *t1 = jl_fieldref_noalloc(v, 1);
-        v = jl_fieldref(v, 0);
+        jl_value_t *t1 = jl_fieldref(v, 1);
         if (jl_is_symbol(t1))
             f_lib = jl_symbol_name((jl_sym_t*)t1);
         else if (jl_is_string(t1))
             f_lib = jl_string_data(t1);
         else
             JL_TYPECHK(cglobal, symbol, t1)
+        v = jl_fieldref(v, 0);
     }
 
     char *f_name = NULL;
```

**Pull request merged**